dango/types/src/vesting.rs

```diff
@@ -0,0 +1,139 @@
+use {
+    anyhow::ensure,
+    grug::{
+        Addr, Coin, Duration, MultiplyFraction, NumberConst, Timestamp, Udec128, Uint128, Undefined,
+    },
+    std::collections::BTreeMap,
+};
+
+pub type ClaimablePosition = Position<Uint128>;
+
+#[grug::derive(Serde)]
+pub struct InstantiateMsg {}
+
+#[grug::derive(Serde)]
+pub enum ExecuteMsg {
+    CreatePosition {
+        user: Addr,
+        schedule: Schedule<Option<Timestamp>>,
+    },
+    Claim {
+        idx: u64,
+    },
+}
+
+#[grug::derive(Serde, QueryRequest)]
+pub enum QueryMsg {
+    #[returns(ClaimablePosition)]
+    Position { idx: u64 },
+    #[returns(BTreeMap<u64, ClaimablePosition>)]
+    Positions {
+        start_after: Option<u64>,
+        limit: Option<u32>,
+    },
+    #[returns(BTreeMap<u64, ClaimablePosition>)]
+    PositionsByUser {
+        user: Addr,
+        start_after: Option<u64>,
+        limit: Option<u32>,
+    },
+}
+
+#[grug::derive(Serde, Borsh)]
+pub struct Schedule<T = Timestamp> {
+    pub start_time: T,
+    pub cliff: Option<Duration>,
+    pub vesting: Option<Duration>,
```

Comment:
I don't see why these two should be optional. The smart contract should follow our legal contract. In our legal contract there is always a cliff and a vesting period. Let's not over complicate things.

Reply:
Ok. The scope of this contract was intended to be used for any vesting purpose, not only for contract agreements (the contract can be used by anyone). This is the reason why both cliff and vesting are options. Anyway having just duration without option allow to set a zero duration to skip a phase